server/src/main/java/org/opensearch/node/resource/tracker/AverageIoUsageTracker.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.node.resource.tracker;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.monitor.fs.FsInfo.DeviceStats;
+import org.opensearch.monitor.fs.FsService;
+import org.opensearch.node.IoUsageStats;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.HashMap;
+
+/**
+ * AverageIoUsageTracker tracks the IO usage by polling the FS Stats for IO metrics every (pollingInterval)
+ * and keeping track of the rolling average over a defined time window (windowDuration).
+ */
+public class AverageIoUsageTracker extends AbstractAverageUsageTracker {
+
+    private static final Logger LOGGER = LogManager.getLogger(AverageIoUsageTracker.class);
+    private final FsService fsService;
+    private final HashMap<String, Long> prevIoTimeDeviceMap;
+    private long prevTimeInMillis;
+    private final IoUsageStats ioUsageStats;
+
+    public AverageIoUsageTracker(FsService fsService, ThreadPool threadPool, TimeValue pollingInterval, TimeValue windowDuration) {
+        super(threadPool, pollingInterval, windowDuration);
+        this.fsService = fsService;
+        this.prevIoTimeDeviceMap = new HashMap<>();
+        this.prevTimeInMillis = -1;
+        this.ioUsageStats = new IoUsageStats(-1);
+    }
+
+    /**
+     * Get current IO usage percentage calculated using fs stats
+     */
+    @Override
+    public long getUsage() {
+        long usage = 0;
+        if (this.preValidateFsStats()) {
+            return usage;
+        }
+        // Currently even during the raid setup we have only one mount device and it is giving 0 io time from /proc/diskstats
+        DeviceStats[] devicesStats = fsService.stats().getIoStats().getDevicesStats();
+        long latestTimeInMillis = fsService.stats().getTimestamp();
+        for (DeviceStats devicesStat : devicesStats) {
+            long devicePreviousIoTime = prevIoTimeDeviceMap.getOrDefault(devicesStat.getDeviceName(), (long) -1);
+            long deviceCurrentIoTime = devicesStat.ioTimeInMillis();
+            if (prevTimeInMillis > 0 && (latestTimeInMillis - this.prevTimeInMillis > 0) && devicePreviousIoTime > 0) {
+                long absIoTime = (deviceCurrentIoTime - devicePreviousIoTime);
+                long deviceCurrentIoUsage = absIoTime * 100 / (latestTimeInMillis - this.prevTimeInMillis);
+                // We are returning the maximum IO Usage for all the attached devices
```

Comment:
Is this right? What if the boot drive does a lot of IO and the EBS devices which are actually doing IO are idle?  Would that be right experience?

Reply:
FS stats only have the details of EBS/data devices. If boot and data drive is same then only boot drive is part of the `iostats` devices list. Also in previous approach we are doing avg of the all the devices which will not give exact impact on each device. Let me know if you have any other thoughts.

Reply:
I think there are some assumptions baked in, that I am not aware of:
1. From the comment, it looks like there is only one logical IO device at all times.  If there is more than one device, we create a logical volume.
2. If there is a separate boot volume, that is not included in the stats? (not sure)

The API is returning a summary statistic, but it is only returning the max of all the devices. The average was a summary statistic.  If you are returning max, should you name the statistic appropriately?

Reply:
Yes, I agree with @kkmr , we can have clear stats name for max usage.

If I understand correctly,  These statistics are only for data volume, and when there are multiple data volumes, we will only return device stats where usage is maximum. Because, as @ajaymovva said,
```Also in previous approach we are doing avg of the all the devices which will not give exact impact on each device```.

Reply:
Yes will update stats name accordingly.

Reply:
Updated